dev/src/index.ts

```diff
@@ -1198,15 +1219,100 @@ export class Firestore implements firestore.Firestore {
     this.bulkWritersCount -= 1;
   }
 
+  /**
+   * Recursively deletes all documents and subcollections at and under the
+   * specified level.
+   *
+   * If any deletes fail, the promise is rejected with an error message
+   * containing the number of failed writes and the stack trace of the last
+   * failed write. The provided reference is deleted regardless of whether
+   * all deletes succeeded, except when Firestore fails to fetch the provided
+   * reference's descendants.
+   *
+   * Firestore uses a BulkWriter instance with default settings to perform the
+   * deletes. To customize throttling rates or add success/error callbacks,
+   * pass in a custom BulkWriter instance.
+   *
+   * @param ref The reference of a document or collection to delete.
+   * @param bulkWriter Custom BulkWriter instance with which to perform the
+   * deletes with.
+   * @return A promise that resolves when all deletes have been performed.
+   * The promise is rejected if any of the deletes fail.
+   */
+  recursiveDelete<T = DocumentData>(
+    ref: CollectionReference<T> | DocumentReference<T>,
+    bulkWriter?: BulkWriter
+  ): Promise<void> {
+    const docStream = this.getAllDescendants(ref);
+    const writer = bulkWriter ?? this.getBulkWriter();
+    const deleteCompleted = new Deferred<void>();
+    let errorCount = 0;
+    let lastError: Error | undefined;
+
+    // Capture the error stack to preserve stack tracing across async calls.
+    const stack = Error().stack!;
+
+    docStream
+      .on('error', err => {
+        err.code = Status.FAILED_PRECONDITION;
+        err.message =
+          'Failed to fetch children documents. ' +
+          'The provided reference was not deleted.';
+        deleteCompleted.reject(wrapError(err, stack));
```

Comment:
Is not deleting the provided reference in the case of a failed StructuredQuery the ideal, expected behavior? I need a quick sanity check. Here are the options:
1. Do not delete the provided reference if StructuredQuery stream ever errors. 
2. Do not delete provided reference iff StructuredQuery stream errors before any deletes are enqueued.
3. Delete the reference regardless of what happens.

I think we originally agreed on Option 2. It makes logical sense -- if Firestore can't fetch any of the descendants, that's a failed precondition, so Firestore shouldn't delete the provided reference. However, what's tripping me up is if the StructuredQuery stream errors out midway through the deletion process. In this case, an argument can be made both ways -- Firestore shouldn't delete the provided reference because it couldn't fetch the descendants (failed precondition argument) OR Firestore should delete the provided reference to be consistent in behavior (the provided reference is deleted even if BulkWriter fails to delete some references). 

Trying to document all these edge cases led me to consider Option 3 -- if I'm a developer who simply wants the collection or document deleted, I don't care about the provided reference, and Firestore should just nuke whatever it can get a hold of.

Reply:
The query can error at any point in time. I would suggest we make no guarantees whatsoever, which makes the overall behavior of the API much more predictable. If any operation fails, the database tree is in an undefined state, and the user needs to try again.

Reply:
Done. Changed it to always delete the provided reference when possible, but kept the error message in for query stream failures. Not sure if we want the FAILED_PRECONDITION status code though.